bin/packages/check-build-type-declaration-files.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * The purpose of this script is to double-check certain published build types.
+ * When we translate JSDoc into types that we wish to publish, we occasionally
+ * disable the type checker for JS files. Unfortunately, this can result in invalid
+ * index.d.ts files being published.
+ *
+ * This script verifies the publishd index.d.ts file for every package which both
+ * builds types and also sets checkJs to false in its tsconfig.json. This is done
+ * by running `tsc --noEmit` on the $package/build-types/index.d.ts file. This also
+ * verifies everything index.d.ts references, which essentially checks the entire
+ * public api of the type declarations for that package.
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/49650 for more discussion.
+ */
+
+/**
+ * External dependencies
+ */
+const fs = require( 'fs' ).promises;
+const path = require( 'path' );
+const { exec } = require( 'child_process' );
+const chalk = require( 'chalk' );
+
+// Tsconfig can be JSON5, which accepts comments, which need to be stripped so we can parse the JSON.
+function stripJsonComments( jsonString ) {
+	const lines = jsonString.split( '\n' );
+	const cleanedLines = lines.map( ( line ) => line.replace( /\/\/.*$/, '' ) );
+	return cleanedLines.join( '\n' );
+}
+
+async function checkDeclarationFile( file ) {
+	return new Promise( ( resolve, reject ) => {
+		exec( `npx tsc --noEmit ${ file }`, ( error, stdout, stderr ) => {
```

Comment:
We don't use `npx` elsewhere in the repo, but it is automatically installed with npm version 6, and uses the resolved `tsc` version from `node_modules`.

Reply:
I believe that's totally fine and standard.
the only downside here is that we're spawning a new shell for each file by using `exec`, but that's fine too. I think it will also divide the work onto multiple CPU cores as a side-effect.

Reply:
> but that's fine too. I think it will also divide the work onto multiple CPU cores as a side-effect.

Yep, that was my intention for this portion, which is probably the most resource intensive part of the script. The nice thing at this point is that the compiled declaration files shouldn't have any uncompiled dependencies, so we don't need to be smart about the order in which we run things.